File: voice-router/email.go
package main

import (
	"fmt"
	"log/slog"
	"net/smtp"
	"strings"
	"time"
)

// sendVoicemailEmail emails a voicemail notification with a link to the recording.
func sendVoicemailEmail(to, callerNumber, recordingURL string, receivedAt time.Time) error {
	subject := fmt.Sprintf("New Voicemail from %s", formatCallerNumber(callerNumber))

	body := buildEmailBody(callerNumber, recordingURL, receivedAt)

	msg := buildMIMEMessage(cfg.SMTPFromAddr, cfg.SMTPFromName, to, subject, body)

	addr := fmt.Sprintf("%s:%s", cfg.SMTPHost, cfg.SMTPPort)
	auth := smtp.PlainAuth("", cfg.SMTPUser, cfg.SMTPPass, cfg.SMTPHost)

	slog.Info("Sending voicemail email", "smtp", addr, "from", cfg.SMTPFromAddr, "to", to)

	if err := smtp.SendMail(addr, auth, cfg.SMTPFromAddr, []string{to}, []byte(msg)); err != nil {
		return fmt.Errorf("smtp.SendMail: %w", err)
	}
	return nil
}

func buildEmailBody(callerNumber, recordingURL string, receivedAt time.Time) string {
	est, _ := time.LoadLocation("America/New_York")
	formatted := receivedAt.In(est).Format("Monday, January 2, 2006 at 3:04 PM MST")

	return fmt.Sprintf(`You have received a new voicemail.

Caller:     %s
Received:   %s

Recording:
%s

--
Store Phone System
This is an automated notification. Do not reply to this email.
`, formatCallerNumber(callerNumber), formatted, recordingURL)
}

func buildMIMEMessage(fromAddr, fromName, to, subject, body string) string {
	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("From: %s <%s>\r\n", fromName, fromAddr))
	sb.WriteString(fmt.Sprintf("To: %s\r\n", to))
	sb.WriteString(fmt.Sprintf("Subject: %s\r\n", subject))
	sb.WriteString("MIME-Version: 1.0\r\n")
	sb.WriteString("Content-Type: text/plain; charset=UTF-8\r\n")
	sb.WriteString("\r\n")
	sb.WriteString(body)
	return sb.String()
}

func formatCallerNumber(number string) string {
	// Basic formatting: +14073082412 -> (407) 308-2412
	digits := strings.Map(func(r rune) rune {
		if r >= '0' && r <= '9' {
			return r
		}
		return -1
	}, number)

	if len(digits) == 11 && digits[0] == '1' {
		digits = digits[1:]
	}
	if len(digits) == 10 {
		return fmt.Sprintf("(%s) %s-%s", digits[:3], digits[3:6], digits[6:])
	}
	return number
}

File: telephony/src/lib/email.ts
// src/lib/email.ts
import nodemailer from "nodemailer";
import { getConfig } from "./config";

function formatCallerNumber(number: string): string {
  const digits = number.replace(/\D/g, "");
  const d = digits.length === 11 && digits[0] === "1" ? digits.slice(1) : digits;
  if (d.length === 10) return `(${d.slice(0, 3)}) ${d.slice(3, 6)}-${d.slice(6)}`;
  return number;
}

export async function sendVoicemailEmail(
  to: string,
  callerNumber: string,
  recordingUrl: string,
  receivedAt: Date
): Promise<void> {
  const cfg = getConfig();
  const formatted = formatCallerNumber(callerNumber);
  const dateStr = receivedAt.toLocaleDateString("en-US", {
    weekday: "long",
    year: "numeric",
    month: "long",
    day: "numeric",
    hour: "numeric",
    minute: "2-digit",
    timeZoneName: "short",
    timeZone: "America/New_York",
  });

  const transport = nodemailer.createTransport({
    host: cfg.smtpHost,
    port: parseInt(cfg.smtpPort),
    auth: { user: cfg.smtpUser, pass: cfg.smtpPass },
  });

  await transport.sendMail({
    from: `"${cfg.smtpFromName}" <${cfg.smtpFromAddr}>`,
    to,
    subject: `New Voicemail from ${formatted}`,
    text: `You have received a new voicemail.\n\nCaller:     ${formatted}\nReceived:   ${dateStr}\n\nRecording:\n${recordingUrl}\n\n--\nJah and Co Phone System\nThis is an automated notification. Do not reply to this email.`,
  });
  console.log(`[email] voicemail sent to ${to}`);
}
